addons/website/static/src/js/tours/tour_utils.js

```diff
@@ -319,7 +319,8 @@ function clickOnExtraMenuItem(stepOptions, backend = false) {
         trigger: `${backend ? "iframe" : ""} #top_menu`,
         run: function () {
             const extraMenuButton = this.$anchor[0].querySelector('.o_extra_menu_items a.nav-link');
-            if (extraMenuButton) {
+            // Don't click on the extra menu button if it's already visible.
+            if (extraMenuButton && !extraMenuButton.classList.contains("show")) {
```

Comment:
@xO-Tx I don't like this kind of stuff in tests, I'd prefer to test for a consistent behavior ("it's supposed to be shown / not shown", not "the test should work if shown or not"). That's true we already check for the extraMenuButton existence here, which I don't like either (in master, we should change that). But in your case, you are just hiding a race condition (not the fact that they may be a lot of menu or not): the test passes without this change... most of the time. Although your commit does not explain that... maybe I am misunderstanding something?

Reply:
@qsm-odoo actually, I fixed it that way because I noticed in the build where it fails that only the "Shop" menu item was available in the navbar (see:  [the screenshot before failure](https://runbot.odoo.com/runbot/build/58139118)).
So before the `clickOnExtraMenuItem()` step, the menu had the following structure:

`["Home" "Shop" "Contact Us" "Mega Menu (added by the tour)" "menu item (added by the tour)"]`

Which explains why the "+" dropdown didn't appear until the editor's sidebar was added (This case should be handled by `clickOnExtraMenuItem()` anyway if the test was done on a database with few apps):

```text
1- [menu_item_0 menu_item_1 menu_item_2...] (no extra menu)
2- Switch to "edit" mode > sidebar added > viewport resized.
3- [menu_item_0 menu_item_1... [+]] (extra menu added the first time after resize, so it will be
closed by default and clickOnExtraMenuItem() needs to force the click()).
```

https://github.com/odoo/odoo/assets/25032876/9451973d-bd6d-4d22-ba41-d9c6fc870130

Do you know exactly why these tests were done using only the `website_sale` menu item :thinking:? Because the tour is not changing the navbar structure before the failure, it only adds a mega menu and a menu item ?

Reply:
> Do you know exactly why these tests were done using only the website_sale menu item 🤔?

Runbot builds are not made always with all the apps. Anyone can ask to make a build with any subset of apps, and tests should work in all.

> Which explains why the "+" dropdown didn't appear until the editor's sidebar was added

Yep, I understand the case of the "+" menu appearing but still don't understand why your chance in the util is necessary. The "+" menu appearing is already covered by the existing `if`. Why is checking "if not already opened" necessary? Can't we go with only restoring the tour step?

Note that I made some tours a while ago that tested all cases of extra menu (see `test_website_extra_items_no_dirty_page`):
- no + at all
- no + outside edit mode, + when entering
- + in both cases
=> At the start of the test, I remove all menus and configure the menus as I want them to be. Not to do here, but in master I think it would be good that all tests that rely on clicking on menu items that are not at the start do that.

Reply:
> Yep, I understand the case of the "+" menu appearing but still don't understand why your change in
the util is necessary. The "+" menu appearing is already covered by the existing if. Why is checking
"if not already opened" necessary? Can't we go with only restoring the tour step?

Actually, the change in "util" is needed because of the "extra menu auto open" behaviour.
Let's take this scenario **["+" menu outside edit mode & "+" menu also when entering]**:
- The extra menu will be already opened because of this behaviour after entering "Edit" mode.
- The step `clickOnExtraMenuItem` will find the "+" and `click()` to close it :grimacing:. The idea here is to restore the click step to cover the `[no "+" before edit, "+" after edit]` case and add the diff in "util" to cover the `["+" in both cases]` scenario :+1:

> => At the start of the test, in python, I remove all menus and configure the menus as I want them to be.
Not to do here, but in master I think it would be good that all tests that rely on clicking on menu items that are not at the start do that.

Isn't better to not preconfigure this kind of tours? If we already have a predefined menu config, it will be like ignoring other possible scenarios… the extra menu behaviour is affected by multiple factors: menu items, initial viewport size, font size in the navbar, the size of the components added next to the iframe (sidebar)… and fixing the navbar items may prevent discovering issues like the one this PR is fixing :thinking:?

Reply:
@xO-Tx 

> If we already have a predefined menu config, it will be like ignoring other possible scenarios [...] fixing the navbar items may prevent discovering issues like the one

Exactly but IMO it's the goal of a test: testing predefined well-determined behaviors in a predefined well-determined set of conditions. Discovering that different set of conditions creates a bug should lead to more tests with those found conditions, not rely on "this test will maybe some day test that set of conditions by mistake". And in this case, we did not find a bug: we are just changing the test :wink: 

But of course, we can just keep the "if + menu is there" for now. No need in changing everything.

> The extra menu will be already opened because of this behaviour after entering "Edit" mode.

Why? In the case of a test, if you did not click on it outside of edit mode, why would he already be opened in edit mode? I am ok with just re-adding the step and let it click on the "+" button "if it is there" because some testing condition make it so it is there... but I don't think there is any value in "if it is not already opened", this is IMO just hiding a potential bug, not being robust to different testing condition.

Reply:
> Why? In the case of a test, if you did not click on it outside of edit mode, why would he already be opened in edit mode?

@qsm-odoo [It's already clicked before going to edit mode](https://github.com/odoo/odoo/pull/155559/commits/b5902e2c61908361495e424a1fc6b94bd799906b#diff-7414ebbca8165cd210cf311d7e5d60438a60f95247369c2ec89f9f0e155e9c06L109), which makes the tour fail with the "restored" click step (maybe I missed something here but it makes sense IMO):

A- **[no "+" before edit, "+" after edit]**: The click step before edit mode won't do anything, and we need to restore the click step after "edit mode switch" to open the "+" dropdown.

B- **["+" before and after edit]**: The click step before edit mode will open the dropdown, and after the "switch to edit mode" resize, the code will simply do the `_adapt()` and show the dropdown again, and the "click" step added in [A] will hide it instead of displaying it :thinking: